File: src/mappings/language.rs
pub fn get<'a>(
    filetype: &'a str,
    filename: &str,
) -> Option<(&'a str, &'a str)> {
    let language = match filetype {
        "Cord.new" => ("text", "New file"),
        "Cord.unknown" => ("text", filetype),
        "autohotkey" => ("ahk", "AutoHotkey"),
        "asm" => ("assembly", "Assembly"),
        "sh" => ("shell", "Shell script"),
        "c" => ("c", "C "),
        "clojure" => ("clojure", "Clojure"),
        "cpp" => ("cpp", "C++"),
        "cr" => ("crystal", "Crystal"),
        "cs" => ("csharp", "C#"),
        "css" => ("css", "CSS"),
        "d" => ("d", "D "),
        "dart" => ("dart", "Dart"),
        "dockerfile" => ("docker", "Docker"),
        "dosbatch" => ("shell", "Batch"),
        "elixir" => ("elixir", "Elixir"),
        "eelixir" => ("elixir", "Embedded Elixir"),
        "heex" => ("phoenix", "Phoenix"),
        "erlang" => ("erlang", "Erlang"),
        "fsharp" => ("fsharp", "F#"),
        "git" | "gitattributes" | "gitconfig" | "gitignore"
        | "gitsendemail" => ("git", "Git"),
        "go" => ("go", "Go"),
        "groovy" => {
            if filename == "build.gradle" {
                ("gradle", "Gradle")
            } else {
                ("groovy", "Groovy")
            }
        }
        "haskell" => ("haskell", "Haskell"),
        "html" => ("html", "HTML"),
        "java" => ("java", "Java"),
        "javascript" => ("javascript", "JavaScript"),
        "javascriptreact" => ("react", "JSX"),
        "json" => ("json", "JSON"),
        "julia" => ("julia", "Julia"),
        "kotlin" => {
            if filename == "build.gradle.kts" {
                ("gradle", "Gradle")
            } else {
                ("kotlin", "Kotlin")
            }
        }
        "tex" | "texmf" | "plaintex" => ("latex", "LaTeX"),
        "lisp" => ("lisp", "Lisp"),
        "lua" => ("lua", "Lua"),
        "markdown" => {
            if filename.to_lowercase() == "license.md" {
                ("license", "License file")
            } else {
                ("markdown", "Markdown")
            }
        }
        "nim" => ("nim", "Nim"),
        "nix" => ("nix", "Nix"),
        "ocaml" => ("ocaml", "OCaml"),
        "pascal" => ("pascal", "Pascal"),
        "perl" => ("perl", "Perl"),
        "php" => ("php", "PHP"),
        "ps1" => ("powershell", "PowerShell"),
        "python" => ("python", "Python"),
        "r" => ("r", "R "),
        "ruby" => ("ruby", "Ruby"),
        "rust" => ("rust", "Rust"),
        "scala" => ("scala", "Scala"),
        "sass" | "scss" => ("scss", "Sass"),
        "sql" => ("sql", "SQL"),
        "svelte" => ("svelte", "Svelte"),
        "swift" => ("swift", "Swift"),
        "text" => {
            let filename = filename.to_lowercase();
            if filename == "license" || filename == "license.txt" {
                ("license", "License file")
            } else {
                ("text", "Plain Text")
            }
        }
        "toml" => {
            if filename == "Cargo.toml" {
                ("cargo", "Cargo")
            } else {
                ("toml", "TOML")
            }
        }
        "typescript" => ("typescript", "TypeScript"),
        "typescriptreact" => ("react", "TSX"),
        "v" => ("v", "V "),
        "vim" => ("vim", "VimL"),
        "vue" => ("vue", "Vue"),
        "xml" => {
            if filename == "pom.xml" {
                ("maven", "Maven")
            } else {
                ("xml", "XML")
            }
        }
        "yaml" => ("yaml", "YAML"),
        "zig" => ("zig", "Zig"),
        "zsh" => ("shell", "Zsh"),
        _ => match filename.to_lowercase().rsplit_once('.') {
            Some((_, extension)) => match extension {
                "gml" => ("gml", "GameMaker Language"),
                "pcss" | "postcss" => ("postcss", "PostCSS"),
                _ => return None,
            },
            None => return None,
        },
    };

    Some(language)
}

File: lua/cord/mappings/language.lua
local M = {}

M.default_icon = 'text'
local mappings = {
  ['Cord.new'] = { M.default_icon, 'New file' },
  ['Cord.unknown'] = { M.default_icon, 'Unknown' },
  autohotkey = { 'ahk', 'AutoHotkey' },
  asm = { 'assembly', 'Assembly' },
  astro = { 'astro', 'Astro' },
  sh = { 'shell', 'Shell script' },
  c = { 'c', 'C' },
  clojure = { 'clojure', 'Clojure' },
  cpp = { 'cpp', 'C++' },
  cr = { 'crystal', 'Crystal' },
  cs = { 'csharp', 'C#' },
  css = { 'css', 'CSS' },
  cuda = { 'nvidia', 'CUDA' },
  d = { 'd', 'D' },
  dart = { 'dart', 'Dart' },
  dockerfile = { 'docker', 'Docker' },
  dosbatch = { 'shell', 'Batch' },
  elixir = { 'elixir', 'Elixir' },
  eelixir = { 'elixir', 'Embedded Elixir' },
  heex = { 'phoenix', 'Phoenix' },
  erlang = { 'erlang', 'Erlang' },
  fsharp = { 'fsharp', 'F#' },
  git = { 'git', 'Git' },
  gitattributes = { 'git', 'Git' },
  gitconfig = { 'git', 'Git' },
  gitignore = { 'git', 'Git' },
  gitsendemail = { 'git', 'Git' },
  go = { 'go', 'Go' },
  haskell = { 'haskell', 'Haskell' },
  html = { 'html', 'HTML' },
  java = { 'java', 'Java' },
  javascript = { 'javascript', 'JavaScript' },
  javascriptreact = { 'react', 'JSX' },
  json = { 'json', 'JSON' },
  julia = { 'julia', 'Julia' },
  tex = { 'latex', 'LaTeX' },
  texmf = { 'latex', 'LaTeX' },
  plaintex = { 'latex', 'LaTeX' },
  lisp = { 'lisp', 'Lisp' },
  lua = { 'lua', 'Lua' },
  nim = { 'nim', 'Nim' },
  nix = { 'nix', 'Nix' },
  ocaml = { 'ocaml', 'OCaml' },
  pascal = { 'pascal', 'Pascal' },
  perl = { 'perl', 'Perl' },
  php = { 'php', 'PHP' },
  ps1 = { 'powershell', 'PowerShell' },
  python = { 'python', 'Python' },
  quarto = { 'quarto', 'Quarto' },
  r = { 'r', 'R' },
  ruby = { 'ruby', 'Ruby' },
  rust = { 'rust', 'Rust' },
  scala = { 'scala', 'Scala' },
  sass = { 'scss', 'Sass' },
  scss = { 'scss', 'Sass' },
  sql = { 'sql', 'SQL' },
  svelte = { 'svelte', 'Svelte' },
  swift = { 'swift', 'Swift' },
  typescript = { 'typescript', 'TypeScript' },
  typescriptreact = { 'react', 'TSX' },
  v = { 'v', 'V' },
  vim = { 'viml', 'VimL' },
  vue = { 'vue', 'Vue' },
  yaml = { 'yaml', 'YAML' },
  zig = { 'zig', 'Zig' },
  zsh = { 'shell', 'Zsh' },
}

local special_cases = {
  groovy = function(filename)
    return filename == 'build.gradle' and { 'gradle', 'Gradle' }
      or { 'groovy', 'Groovy' }
  end,
  kotlin = function(filename)
    return filename == 'build.gradle.kts' and { 'gradle', 'Gradle' }
      or { 'kotlin', 'Kotlin' }
  end,
  markdown = function(filename)
    return filename:lower() == 'license.md' and { 'license', 'License file' }
      or { 'markdown', 'Markdown' }
  end,
  text = function(filename)
    local name = filename:lower()
    return (name == 'license' or name == 'license.txt')
        and { 'license', 'License file' }
      or { M.default_icon, 'Plain Text' }
  end,
  toml = function(filename)
    return filename == 'Cargo.toml' and { 'cargo', 'Cargo' }
      or { 'toml', 'TOML' }
  end,
  xml = function(filename)
    return filename == 'pom.xml' and { 'maven', 'Maven' } or { 'xml', 'XML' }
  end,
}

local extension_mappings = {
  gml = { 'gml', 'GameMaker Language' },
  hx = { 'haxe', 'Haxe' },
  hxml = { 'haxe', 'Haxe' },
  pcss = { 'postcss', 'PostCSS' },
  postcss = { 'postcss', 'PostCSS' },
}

M.get = function(filetype, filename)
  if special_cases[filetype] then return special_cases[filetype](filename) end
  if mappings[filetype] then return mappings[filetype] end

  local ext = filename:match '%.([^%.]+)$'
  if ext then return extension_mappings[ext:lower()] end
end

return M
